File: crates/durable-java/src/main/java/cleat/DurableEntry.java
package cleat;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Marks a method as a cleat workflow entry point.
 * <p>
 * Methods annotated with {@code @DurableEntry} must be {@code public static}
 * and take a {@link HostCalls} instance as their first parameter.
 * Additional parameters are deserialized from the workflow input JSON.
 * <p>
 * The annotation processor ({@link DurableEntryProcessor}) generates a
 * WASM export wrapper at compile time for each annotated method.  The
 * generated class implements the cleat ABI export signature:
 * <pre>
 *   (argsPtr: i32, argsLen: i32, outPtr: i32, maxOutLen: i32) -> i64
 * </pre>
 * <p>
 * <strong>Example usage:</strong>
 * <pre>{@code
 * public class MyWorkflow {
 *     @DurableEntry(name = "place_order")
 *     public static String placeOrder(HostCalls h, String input) {
 *         h.durableLog("Processing order");
 *         String result = h.durableCall("orders", "create", input).getValue();
 *         return result;
 *     }
 * }
 * }</pre>
 *
 * @see HostCalls
 * @see DurableEntryProcessor
 */
@Retention(RetentionPolicy.SOURCE)
@Target(ElementType.METHOD)
public @interface DurableEntry {

    /**
     * The WASM export name for this workflow entry point.
     * Defaults to the Java method name if left empty.
     */
    String name() default "";
}

File: packages/cleat-as/assembly/durable-entry.ts
/**
 * Marker decorator for workflow entry points.
 *
 * Identifies a function as a cleat workflow entry point. The actual WASM
 * export generation with the ABI-compatible signature
 * `(argsPtr: i32, argsLen: i32, outPtr: i32, maxOutLen: i32) -> i64` is
 * handled by the cleat-as-transformer plugin at compile time.
 *
 * Usage:
 * ```ts
 * &#64;durableEntry("PlaceOrder")
 * export function placeOrder(argsPtr: i32, argsLen: i32, outPtr: i32, maxOutLen: i32): i64 {
 *   // workflow logic
 * }
 * ```
 *
 * @param name - Optional workflow name. Defaults to the function name.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function durableEntry(name: string = ""): (target: usize, propertyKey: string, descriptor: usize) => void {
  return function (
    _target: usize,
    _propertyKey: string,
    _descriptor: usize,
  ): void {
    // Marker decorator — no runtime behavior.
    // The transformer plugin reads this decorator metadata at compile time
    // to generate the appropriate WASM export with the ABI signature.
  };
}
